api/s2n.h

```diff
@@ -122,12 +123,49 @@ extern int s2n_connection_wipe(struct s2n_connection *conn);
 extern int s2n_connection_free(struct s2n_connection *conn);
 extern int s2n_shutdown(struct s2n_connection *conn, s2n_blocked_status *blocked);
 
+typedef enum { S2N_CERT_AUTH_NONE, S2N_CERT_AUTH_REQUIRED } s2n_cert_auth_type;
+extern int s2n_config_get_client_auth_type(struct s2n_config *config, s2n_cert_auth_type *client_auth_type);
+extern int s2n_config_set_client_auth_type(struct s2n_config *config, s2n_cert_auth_type client_auth_type);
+extern int s2n_connection_get_client_auth_type(struct s2n_connection *conn, s2n_cert_auth_type *client_auth_type);
+extern int s2n_connection_set_client_auth_type(struct s2n_connection *conn, s2n_cert_auth_type client_auth_type);
+extern int s2n_connection_get_client_cert_chain(struct s2n_connection *conn, uint8_t **der_cert_chain_out, uint32_t *cert_chain_len);
+
```

Comment:
Yep, my thinking was that we could keep them generic so they could be reused for both server and client auth settings. This will only be running/configured in either server or client mode, not both at the same time. If you were configuring an s2n server and set the auth type, it would be for the client cert. If you were running s2n in client mode, you would be setting the auth type for the server. I think of it as setting the "peer" auth type, where "peer" would mean different sides of the connection depending on which mode you were in.

Since both client and server do very similar operations to authenticate each other, it would cut down on the number of mode-specific functions we would have to have in our API (which has tended to favor "mode-agnostic" functions so far).